### DISCUSSION TITLE
Problem with Energy.Active.Import.Register since 03.02.2020 (V0.4.15 - V0.4.16)

### DISCUSSION THREAD

**Discussion by McFranky** [lbbrhzn/ocpp#366] [General]
ABB Terra Firmware 1.5.2

Problem:
Energy.Active.Import.Register 
I got dropouts over the time
![grafik](https://user-images.githubusercontent.com/59682922/152502268-03344331-ab9d-4eb5-a4a4-85bc0d204c30.png)

The problem can be identifed in the log:
Normal MeterValues; Energy.Active.Import.Register = 895Wh = 0,89kWh; thats correct
2022-02-04 06:29:19 INFO (MainThread) [ocpp] ILET01: receive message [2, "5430689", "MeterValues", {"connectorId": 1, "transactionId": 1643951526, "meterValue": [{"timestamp": "2022-02-04T05:29:17.000Z", "sampledValue": [{"value": "222.30", "context": "Sample.Periodic", "format": "Raw", "measurand": "Voltage", "phase": "L1-N", "unit": "V"}, {"value": "14.30", "context": "Sample.Periodic", "format": "Raw", "measurand": "Current.Import", "phase": "L1", "unit": "A"}, {"value": "3144", "context": "Sample.Periodic", "format": "Raw", "measurand": "Power.Active.Import", "phase": "L1", "unit": "W"}, {"value": "228.70", "context": "Sample.Periodic", "format": "Raw", "measurand": "Voltage", "phase": "L2-N", "unit": "V"}, {"value": "0.0", "context": "Sample.Periodic", "format": "Raw", "measurand": "Current.Import", "phase": "L2", "unit": "A"}, {"value": "0", "context": "Sample.Periodic", "format": "Raw", "measurand": "Power.Active.Import", "phase": "L2", "unit": "W"}, {"value": "226.40", "context": "Sample.Periodic", "format": "Raw", "measurand": "Voltage", "phase": "L3-N", "unit": "V"}, {"value": "0.0", "context": "Sample.Periodic", "format": "Raw", "measurand": "Current.Import", "phase": "L3", "unit": "A"}, {"value": "0", "context": "Sample.Periodic", "format": "Raw", "measurand": "Power.Active.Import", "phase": "L3", "unit": "W"}, {"value": "**895**", "context": "Sample.Periodic", "format": "Raw", "measurand": "Energy.Active.Import.Register", "unit": "Wh"}, {"value": "16", "context": "Sample.Periodic", "format": "Raw", "measurand": "Current.Offered", "unit": "A"}]}]}]

A litle bit later than this:
2022-02-04 06:30:01 INFO (MainThread) [ocpp] ILET01: receive message [2, "6287334", "MeterValues", {"connectorId": 1, "meterValue": [{"timestamp": "2022-02-04T05:30:00.000Z", "sampledValue": [{"value": "**10531**", "context": "Sample.Clock", "format": "Raw", "measurand": "Energy.Active.Import.Register", "location": "Inlet"}]}]}]

UnNormal MeterValues; Energy.Active.Import.Register = 10531Wh = 10,53kWh; thats not correct (Is this the total energy (not session)?)

The second is without transactionID! 

I have a new Firmware on the ABB Terra. I think thats a new problem. Can you filter only Energy.Active.Import.Register values with transactionID?

Greetings from Berlin
Frank

**Comment by drc38** [comment]
If you reinstall the integration and set "Charger idle sampling interval" to 0 that will turn off the clock sampling as a temporary fix.

**Reply by lbbrhzn** [reply]
@drc38 do we need to distinguish between transaction / connector related meter values and overall meter values in the handler of the meter values, like we did for the status/error?

**Reply by drc38** [reply]
@McFranky with the clock sampling turned off is Energy.Active.Import.Register value equal to the Energy.Session value?

**Reply by drc38** [reply]
> @drc38 do we need to distinguish between transaction / connector related meter values and overall meter values in the handler of the meter values, like we did for the status/error?

Not sure how we would, seems it is manufacturer dependent! That is they have interpreted the ocpp 1.6 spec differently. My charger always reports an absolute meter value. What does the wallbox do?

**Reply by McFranky** [reply]
Yes, Energy.Active.Import.Register value is equal to the Energy.Session value, if there is no sampling.
I think, that the new firmware on ABB Terra sends total energy on sampling and session energy while session is on.

**Reply by drc38** [reply]
I think the simplest approach is to assume `Energy.Active.Import.Register` should be monotonically increasing (ie majority of chargers output). An if statement can be used to check if new value > old value and if not, updated value = new delta + old

**Reply by lbbrhzn** [reply]
> Not sure how we would, seems it is manufacturer dependent! That is they have interpreted the ocpp 1.6 spec differently. My charger always reports an absolute meter value. What does the wallbox do?

MeterValues for connect_id=0 should be for all connectors combined, MeterValues with connector_id>0 should be for a specific connector. Not sure how many chargers there are for domestic use with multiple connectors though.

**Reply by drc38** [reply]
I don't think it will output MeterValues for connector_id=0 when charging, @McFranky can you check?

**Reply by lbbrhzn** [reply]
From OCPP spec:

MeterValues_req,connecor_id:
The id of the Connector from which samples were taken. If the connectorId is 0, it is associated with the entire Charge Point. If the connectorId is 0 and the Measurand is energy related, the sample SHOULD be taken from the main energy meter.

**Reply by lbbrhzn** [reply]
@drc38: We should distinguish between meter values with transaction_id and without.

_The Charge Point SHALL send a MeterValues.req PDU for offloading meter values. The request PDU SHALL contain for each sample:
The transaction Id of the transaction to which these values are related, if applicable. If there is no transaction in progress or if the values are taken from the main meter, then transaction id may be omitted._

For each energy sensor that we create we should decide whether it displays main or transaction related data. If they are transaction related they should be reset by the charger at the start of a transaction, and otherwise not. In our current implementation we ignore the transaction_id, and assume that all meter values are **not** transaction related. To compute the session energy, we subtract the value at start. 

For non-energy related values (power, current, etc) there should only be differences between main and transaction if it supports multiple connectors, so it supports concurrent transactions. Let's not worry about that now.

**Comment by lbbrhzn** [comment]
I've created issue #374 for this

**Comment by gytisgreitai** [comment]
Hey, should this be fixed? I somehow have a strange graph? 
![image](https://user-images.githubusercontent.com/1690228/198847888-428e5d29-ea96-430e-b8d7-fa0cadc993bf.png)

And the value ... 37690 kwh? is there some decimal issue?

**Reply by drc38** [reply]
The energy import register accumulates from charger installation every charge session's energy. So no issue with its value unless it does not match what is being reported in the log?

**Reply by gytisgreitai** [reply]
So today it seems it dropped from 37 thousand kwh to 6kwh (it's currently charging, so it matches the charger.Energy.Session

![image](https://user-images.githubusercontent.com/1690228/198864806-6a42d83f-6fa3-4f62-9776-d5f992926b76.png)

The charge station - ABB terra is brand new, installed just this month, and the app totals are 16.15kwh 

![image](https://user-images.githubusercontent.com/1690228/198864911-04d8148b-0e54-4c6b-a045-94461a4c78ad.png)

**Reply by drc38** [reply]
Did you set "Charger idle sampling interval" to 0 as suggested above?

**Reply by gytisgreitai** [reply]
sorry, missed that - thought that it was needed before the fix. Removed and re-added, set to zero, let's see how it goes. Also I somehow failed to remove OCPP via UI, had to do that through various .store/core.* files

**Reply by gytisgreitai** [reply]
>The energy import register accumulates from charger installation every charge session's energy

ok, so with sampling interval set to 0 for ABB Terra this does not seem to be the case? Looks like it's equal to sessions energy. Is this yet another quirk of the ABB Terra?

![image](https://user-images.githubusercontent.com/1690228/199920332-b92bb7c4-6eca-4694-8ead-5ee01d433212.png)

**Reply by drc38** [reply]
Probably! If you want it to show the accumulated register instead you could try not selecting the energy.active.import.register when setting up the integration and turn back on the sampling interval.

**Reply by drc38** [reply]
This should now be fixed in release v0.4.42 thanks to @nlindn
